feat(stock): Pick List Scan

dj12djdjs opened the pull request:
closes: https://github.com/frappe/erpnext/issues/30821

### Proposed Solution
Please note for the examples bellow there are 4 items Item A - Item D with barcode 'a' corresponding to Item A, etc...

- [x] Restrict `BarcodeScanner` from creating new rows.
 By default `BarcodeScanner` will create a row if batch_no or item_code doesn't exist.
 In the case of pick list only the rows generated from sales order or other supported documents
 should be considered when scanning. Ex. We are trying to add Item B through barcode 'b' but it doesn't add row
 because Item B isn't in the list of items to pick.
![feat-pick-list-scan_prevent-new-row](https://user-images.githubusercontent.com/29856401/165668039-d8cd3ee0-27d1-42ca-961e-ce8f463b9ca6.gif)


- [x] Restrict `BarcodeScanner` from exceeding maximum pickable qty provided by pick list.
  By default `BarcodeScanner` will continue to increment the quantity of the scanned item.
  In the case of pick list a maximum quantity will be reached. I've modified the `BarcodeScanner` opts to support this.
![feat-pick-list-scan_max-qty](https://user-images.githubusercontent.com/29856401/165667844-1d81adf7-f38b-4019-8a19-2533cf59d804.gif)



- [x] Support multiple rows with the same item code.
![feat-pick-list-scan_multiple-item_codes](https://user-images.githubusercontent.com/29856401/165667378-bdf463ad-f463-4a3f-8fa7-a974a02150ef.gif)


- [x] Prevent Picked Qty from being fulfilled automatically when pick list is submitted. There should be some validation Ex: Picker didn't pick all the items.
![dont allow delivery note with full qty picked](https://user-images.githubusercontent.com/29856401/166480598-4e0b8fef-1a16-4708-8f46-1f8398359268.gif)


- [x] Might be good to have an option prompt for qty after a scan. This way if a picking order calls for 50pcs It won't be necessary to scan all 50pcs. Just one, then input qty to increment. 
![prompt qty](https://user-images.githubusercontent.com/29856401/166952549-3f0fcd26-ddae-4586-8721-0adb146182bf.gif)


**Please post your suggestions / feedback.**

ankush committed:
fix: auto-fulfill picking list when not in scan mode

dj12djdjs committed:
feat: pick list scan fields

dj12djdjs committed:
fix: get correct row to modify with duplicate item_codes and max_qty fields are present

dj12djdjs committed:
fix: show alert when maximum qty scanned is reached

dj12djdjs committed:
feat: increment picked_qty on scan_barcode

ankush commented on erpnext/stock/doctype/pick_list/pick_list.json:
```diff
@@ -126,11 +130,32 @@
+  },
+  {
+   "default": "0",
+   "fieldname": "scan_mode",
```
I don't think this is required. We can always show the scan_barcode field (which has global setting (in stock settings) to hide it if required 😄 


This was required in stock reconciliation because there the default behaviour and scan behaviour were quite conflicting in nature. Here there are no triggers on picked qty.

ankush commented:
>  Restrict BarcodeScanner from creating new rows.

Makes sense for picklist. I was gonna add this too. 


>  Restrict BarcodeScanner from exceeding the maximum pickable qty provided by picklist.

This also makes sense. However, IMO this check should be implemented in Pick List's JS and not in the barcode scanner? Do you know anywhere else where max qty validation will be required with barcode scanning?


>  Support multiple rows with the same item code.

Behavior shown in GIF looks good 👍 


>  Prevent Picked Qty from being fulfilled automatically when pick list is submitted. There should be some validation Ex: Picker didn't pick all the items.


Yeah, this will be a good addition. We can just check before_submit if all stock_qty == picked_qty and if not we can warn the user. 


>  Support Simultaneous pickers scanning items for pick list

Not sure if I understand what this means 😅 

Two pickers working on same doc on different computers? I highly doubt that's doable because of the way the framework works right now.

ankush commented on erpnext/stock/doctype/pick_list_item/pick_list_item.json:
```diff
@@ -187,11 +188,16 @@
    "read_only": 1
+  },
+  {
+   "fieldname": "barcode",
```
This isn't required IMO. 

This was the legacy way to scan barcodes. (Open row > click on barcode field and scan it)

dj12djdjs replied:
Okay, I will remove this. I only added it because I noticed it was in all other doctypes using barcode.

dj12djdjs commented:
> > Support Simultaneous pickers scanning items for pick list
> 
> Not sure if I understand what this means 😅
> 
> Two pickers working on same doc on different computers? I highly doubt that's doable because of the way the framework works right now.

I kind of figured this. It is one of the requirements I have for my client. I will have to develop some middle man to handle this.

szufisher commented:
please also consider catch and handle scan same serial number multi times, currently duplicate serial number will be added to serial no field and quantity increase by 1. program need to prevent this and feedback as warning.

ankush commented:
> I kind of figured this. It is one of the requirements I have for my client. I will have to develop some middle man to handle this.

You can use real-time updates using socket.io for this. Real-time synchronization of normal data structures like JSON (document object) would quite likely lead to conflicts and unexpected behavior 😄 You'll have to implement some kind of locking/synchronization primitives on top of the document object to prevent races. 


> please also consider catch and handle scan same serial number multi times, currently duplicate serial number will be added to serial no field and quantity increase by 1. program need to prevent this and feedback as warning.

This is fixed in develop version. Since it was part of bigger changes it didn't get backported, maybe I should port fixes backto v13 😅

dj12djdjs committed:
revert: barcode scan field from pick list item.

dj12djdjs committed:
revert: scan_mode flag

dj12djdjs commented:
> This also makes sense. However, IMO this check should be implemented in Pick List's JS and not in the barcode scanner? Do you know anywhere else where max qty validation will be required with barcode scanning?

@ankush I agree with what you're saying. I was originally implementing with that concept in mind, until I came to the realization I needed to be able to scan multiple rows of the same item code. The barcode scan API now relies specifying a maximum quantity so it knows to move to the next item.
https://github.com/dj12djdjs/erpnext/blob/8053f2dbcd835602d86dca520b8629b0ac866656/erpnext/public/js/utils/barcode_scanner.js#L199-L202

dj12djdjs committed:
Merge branch 'develop' into feat-picklist-scan

dj12djdjs commented:
> > please also consider catch and handle scan same serial number multi times, currently duplicate serial number will be added to serial no field and quantity increase by 1. program need to prevent this and feedback as warning.
> 
> This is fixed in develop version. Since it was part of bigger changes it didn't get backported, maybe I should port fixes backto v13

@szufisher It seems this might be fixed in develop branch. I'm assuming serial_no has something to do with batch qty. I have no experience with this workflow.

I've setup test site, could you check if the behavior is working as expected for serial no.?
https://feat-picklist-scan.frappe.cloud/
demo@example.com
password

dj12djdjs committed:
fix: syntax

dj12djdjs committed:
fix: warn user pick list is not complete instead of auto fulfilling picked_qty.

dj12djdjs committed:
fix: prevent user from proceeding without all qty picked.

dj12djdjs committed:
fix: linter

dj12djdjs committed:
Merge branch 'feat-picklist-scan' of github.com:dj12djdjs/erpnext into feat-picklist-scan

ankush commented:
> @ankush I agree with what you're saying. I was originally implementing with that concept in mind, until I came to the realization I needed to be able to scan multiple rows of the same item code. The barcode scan API now relies specifying a maximum quantity so it knows to move to the next item.


makes sense. Let's go with max_qty for now, if I can think of a better solution I'll comment / implement later.

dj12djdjs commented:
> > @ankush I agree with what you're saying. I was originally implementing with that concept in mind, until I came to the realization I needed to be able to scan multiple rows of the same item code. The barcode scan API now relies specifying a maximum quantity so it knows to move to the next item.
> 
> makes sense. Let's go with max_qty for now, if I can think of a better solution I'll comment / implement later.

@ankush I'm going to be implementing barcode scan for Packing Slip as well. Maybe max_qty will be useful over there.

dj12djdjs committed:
Merge branch 'develop' into feat-picklist-scan

dj12djdjs committed:
fix(test): manually select picked_qty since before_submit validates qty instead of auto filling it

dj12djdjs committed:
Merge branch 'feat-picklist-scan' of /home/devin/repos/erpnext into feat-picklist-scan

ankush commented on erpnext/public/js/utils/barcode_scanner.js:
```diff
@@ -10,6 +10,11 @@ erpnext.utils.BarcodeScanner = class BarcodeScanner {
 		this.batch_no_field = opts.batch_no_field || "batch_no";
 		this.qty_field = opts.qty_field || "qty";
+		this.max_qty_field = opts.max_qty_field || null;
+		this.allow_new_row = opts.allow_new_row;
+		if (this.allow_new_row === undefined || this.allow_new_row === null) {
+			this.allow_new_row = true;
+		}
```
Using negation of this is simpler. 

`this.dont_allow_new_row` which by default is correct `undefined`.

ankush replied:
alternatively: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

dj12djdjs replied:
Resolved by : https://github.com/frappe/erpnext/pull/30832/commits/7ae89dedd59815cc0dea63a3ecf67b2355ea4425

ankush commented on erpnext/public/js/utils/barcode_scanner.js:
```diff
@@ -72,6 +77,15 @@ erpnext.utils.BarcodeScanner = class BarcodeScanner {
 		if (!row) {
+			if (!this.allow_new_row) {
+				frappe.show_alert({
+					message: __("Maximum quantity scanned for this barcode."),
```
```suggestion
					message: __("Maximum quantity scanned for item {0}.", [item_code]),
```

ankush commented on erpnext/stock/doctype/pick_list/pick_list.py:
```diff
@@ -42,8 +42,13 @@ def before_submit(self):
+						item.idx, item.stock_qty - item.picked_qty, item.stock_uom
+					),
+					title=_("Pick List Incomplete"),
+				)
```
This breaks current behavior for those who don't scan things. They just want to pick 100% qty and submit the pick list.

ankush replied:
we might have to add "scan_mode" afterall to differentiate between users adding qty via scanning vs auto-picking (current behavior)

dj12djdjs replied:
Resolved by https://github.com/frappe/erpnext/pull/30832/commits/d48ab81622de00c1f088caa435468476cfd9c727

ankush commented on erpnext/stock/doctype/pick_list/test_pick_list.py:
```diff
@@ -411,6 +411,8 @@ def test_pick_list_for_items_with_multiple_UOM(self):
 		)
 		pick_list.set_item_locations()
+		for location in pick_list.locations:
+			location.picked_qty = location.stock_qty
```
Modifying tests like this is a red flag 😬  it is indeed breaking an existing workflow.

dj12djdjs replied:
Resolved by https://github.com/frappe/erpnext/pull/30832/commits/4a19c1c19d23bd833d845631a0ddc8a04141a32d

dj12djdjs committed:
Revert "fix(test): manually select picked_qty since before_submit validates qty instead of auto filling it"

This reverts commit c2335ec0

dj12djdjs committed:
fix: max qty message

Co-authored-by: Ankush Menat <ankushmenat@gmail.com>

dj12djdjs committed:
fix: cleanup dont_allow_new_row logic

dj12djdjs committed:
feat: prompt qty on scan

dj12djdjs committed:
fix: cast value to Number type

dj12djdjs committed:
feat: add prompt qty flag to pick list

dj12djdjs commented:
@ankush I've added another feature for prompting qty. It will be useful so user doesn't have to scan an item so many times for high qty. Let me know your thoughts.

ankush commented:
> @ankush I've added another feature for prompting qty. It will be useful so user doesn't have to scan an item so many times for high qty. Let me know your thoughts.


We have also had a similar request, though UX for this needs to be reviewed. Some users would want to scan everything, and some would want a prompt. 

There are even cases where users want prompt for buying cycle (where large qty are bought) but individual scans for the selling cycle where qty is usually small. 😬 



cc: @marination

dj12djdjs commented:
> > @ankush I've added another feature for prompting qty. It will be useful so user doesn't have to scan an item so many times for high qty. Let me know your thoughts.
> 
> We have also had a similar request, though UX for this needs to be reviewed. Some users would want to scan everything, and some would want a prompt.
> 
> There are even cases where users want prompt for buying cycle (where large qty are bought) but individual scans for the selling cycle where qty is usually small. 😬
> 
> cc: @marination

These are also very useful, without these features barcode scan feature is weak. Although, I think I should publish these in a separate PR as they are more for the whole of the barcode scan API.

ankush commented on erpnext/stock/doctype/pick_list/pick_list.json:
```diff
@@ -126,11 +131,38 @@
+  },
+  {
+   "fieldname": "section_break_15",
+   "fieldtype": "Section Break"
```
This section can be hidden when doc is non-draft.

ankush replied:
The alignment of these fields is bit ugly but idk if we can do anything about it, so lets ignore for now. 


<img width="997" alt="Screenshot 2022-05-07 at 5 20 16 PM" src="https://user-images.githubusercontent.com/9079960/167253039-a00e0974-6fc7-42be-9d51-850fd242fc50.png">

ankush replied:
We can't do column breaks like this right now due to the way layout works:

`|--|-|-|`

ankush replied:
test comment

ankush replied:
test

ankush replied:
test

ankush commented on erpnext/public/js/utils/barcode_scanner.js:
```diff
@@ -182,7 +209,11 @@ erpnext.utils.BarcodeScanner = class BarcodeScanner {
+		const existing_item_row = this.items_table.filter((d) => {
+			const [qty, max_qty] = [d[this.qty_field], d[this.max_qty_field] || null];
+			return d.item_code === item_code && ((max_qty === null) || (qty < max_qty));
+		}).splice(0, 1).pop();
```
This max_qty magic is only happening for non-batched items. There's a separate function for fetching row that match item_code and batch_no both, right above this function.

ankush commented on erpnext/stock/doctype/pick_list/pick_list.py:
```diff
@@ -41,8 +41,15 @@ def before_save(self):
-			if item.picked_qty == 0:
+			if self.scan_mode and item.picked_qty < item.stock_qty:
+				frappe.throw(
+					_("Row {0} is short by {1} {2}").format(
```
```suggestion
					_("Row {0} is short by {1} {2}").format(
```

Better wording than "short by" 😅 

Perheps: `Row {} picked qty is less than required qty, additional {pending_qty} {uom} required.`

ankush commented on erpnext/public/js/utils/barcode_scanner.js:
```diff
@@ -106,9 +119,23 @@ erpnext.utils.BarcodeScanner = class BarcodeScanner {
+			};
+
+			if (this.prompt_qty) {
+				frappe.prompt(__("Please enter quantity for item {0}", [item_code]), ({value}) => {
```
Side-effect of this, idk if intentional or not: fractional qty scanning is now also possible 😬

ankush commented on erpnext/stock/doctype/pick_list/pick_list.js:
```diff
@@ -158,6 +158,18 @@ frappe.ui.form.on('Pick List', {
+			prompt_qty: frm.doc.prompt_qty
+		};
+		const barcode_scanner = new erpnext.utils.BarcodeScanner(opts);
+		barcode_scanner.process_scan();
```
Good to see separating out barcode scanner code actually being useful here 😬

ankush commented on erpnext/public/js/utils/barcode_scanner.js:
```diff
@@ -10,6 +10,9 @@ erpnext.utils.BarcodeScanner = class BarcodeScanner {
 		this.qty_field = opts.qty_field || "qty";
+		this.max_qty_field = opts.max_qty_field || null;
+		this.dont_allow_new_row = opts.dont_allow_new_row;
+		this.prompt_qty = opts.prompt_qty;
```
It would be great if we can add basic info about what these fields do. Some are obvious, rest not so much.

ankush reviewed (changes requested):
last round of minor changes. Tested locally, mostly LGTM. 

Parking batch/serial UX for later, it's bit awkward in picklist right now.

ankush committed:
Merge branch 'develop' into feat-picklist-scan

ankush committed:
fix(UX): hide scan section for non-draft docs

ankush committed:
fix: better message for picked qty shortfall

ankush committed:
docs: document barcode_scanner API

ankush committed:
refactor: change alert duration to 3 and modern js

ankush committed:
fix: disable serial no scanning for picklist

Current design of picklist doc doesn't have "picked serial no" so it
doesn't make much sense to scan it. Instead it should increment qty
only. Might add this in future after fixing UX problems

ankush committed:
refactor: single function to fetch related row

There was separate function for batch row which frequently didn't
receive all the love main function received like:
1. empty row reuse
2. max qty validation

Hence it makes sense to combine these in one fat function

ankush commented:
@dj12djdjs please update the docs page. There are some differences here for "prompt" and "scan mode"

ankush committed:
chore: bump ecma version

we transpile to es2017, so allow use of es2020

dj12djdjs commented:
@ankush  Will do, apologies for the delay

dj12djdjs commented:
@ankush Should I target the doc for V14 only?

ankush commented:
@dj12djdjs yes. This depends on changes to barcode scanner that aren't on v13.

dj12djdjs commented:
docs update https://docs.erpnext.com/compare?wiki_page=8c53d48a87&&compare=0fc7ad55c8
